File: herald-admin-typescript/src/namespaces/tenants.ts
import type { HttpTransport } from "../transport.js";
import type { Stream } from "../types.js";

export interface Tenant {
  id: string;
  name: string;
  plan: string;
  key: string;
  created_at: number;
}

export interface TenantWithSecret extends Tenant {
  secret: string;
}

export interface CreateTenantOptions {
  name: string;
  plan?: string;
}

export interface ApiToken {
  token: string;
  scope?: string | null;
}

export class TenantNamespace {
  constructor(private transport: HttpTransport) {}

  async create(opts: CreateTenantOptions): Promise<TenantWithSecret> {
    return this.transport.request<TenantWithSecret>("POST", "/admin/tenants", opts);
  }

  async list(): Promise<Tenant[]> {
    const res = await this.transport.request<{ tenants: Tenant[] }>("GET", "/admin/tenants");
    return res.tenants;
  }

  async get(id: string): Promise<Tenant> {
    return this.transport.request<Tenant>("GET", `/admin/tenants/${encodeURIComponent(id)}`);
  }

  async update(id: string, opts: { name?: string; plan?: string; config?: unknown; event_ttl_days?: number }): Promise<void> {
    await this.transport.request("PATCH", `/admin/tenants/${encodeURIComponent(id)}`, opts);
  }

  async delete(id: string): Promise<void> {
    await this.transport.request("DELETE", `/admin/tenants/${encodeURIComponent(id)}`);
  }

  async createToken(tenantId: string, scope?: string): Promise<ApiToken> {
    const body = scope ? { scope } : undefined;
    return this.transport.request<ApiToken>("POST", `/admin/tenants/${encodeURIComponent(tenantId)}/tokens`, body);
  }

  async listTokens(tenantId: string): Promise<ApiToken[]> {
    const res = await this.transport.request<{ tokens: ApiToken[] }>("GET", `/admin/tenants/${encodeURIComponent(tenantId)}/tokens`);
    return res.tokens;
  }

  async deleteToken(tenantId: string, token: string): Promise<void> {
    await this.transport.request("DELETE", `/admin/tenants/${encodeURIComponent(tenantId)}/tokens/${encodeURIComponent(token)}`);
  }

  async listTenantStreams(tenantId: string): Promise<Stream[]> {
    const res = await this.transport.request<{ streams: Stream[] }>("GET", `/admin/tenants/${encodeURIComponent(tenantId)}/streams`);
    return res.streams;
  }
}

File: herald-admin-ruby/lib/herald_admin/tenants.rb
# frozen_string_literal: true

module HeraldAdmin
  class TenantNamespace
    def initialize(transport)
      @t = transport
    end

    def create(name, plan: nil)
      body = { name: name }
      body[:plan] = plan if plan
      @t.request("POST", "/admin/tenants", body)
    end

    def list
      data = @t.request("GET", "/admin/tenants")
      data["tenants"]
    end

    def get(id)
      @t.request("GET", "/admin/tenants/#{ERB::Util.url_encode(id)}")
    end

    def update(id, name: nil, plan: nil, event_ttl_days: nil)
      body = {}
      body[:name] = name if name
      body[:plan] = plan if plan
      body[:event_ttl_days] = event_ttl_days if event_ttl_days
      @t.request("PATCH", "/admin/tenants/#{ERB::Util.url_encode(id)}", body)
    end

    def delete(id)
      @t.request("DELETE", "/admin/tenants/#{ERB::Util.url_encode(id)}")
    end

    def create_token(tenant_id, scope: nil)
      body = scope ? { scope: scope } : nil
      data = @t.request("POST", "/admin/tenants/#{ERB::Util.url_encode(tenant_id)}/tokens", body)
      data["token"]
    end

    def delete_token(tenant_id, token)
      @t.request("DELETE", "/admin/tenants/#{ERB::Util.url_encode(tenant_id)}/tokens/#{ERB::Util.url_encode(token)}")
    end

    def list_tokens(tenant_id)
      data = @t.request("GET", "/admin/tenants/#{ERB::Util.url_encode(tenant_id)}/tokens")
      data["tokens"]
    end

    def list_streams(tenant_id)
      data = @t.request("GET", "/admin/tenants/#{ERB::Util.url_encode(tenant_id)}/streams")
      data["streams"]
    end
  end
end
